v2/tools/generator/internal/codegen/pipeline/move_types_for_version_migration.go

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rotisserie/eris"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/config"
+)
+
+const MoveTypesForVersionMigrationStageID = "moveTypesForVersionMigration"
+
+// MoveTypesForVersionMigration returns a pipeline stage that moves types into new packages to support
+// migration from legacy versioning (using the v1api prefix) to new versioning (using simple date-based versions).
+// See legacy.go in the astmodel package for more details on the version migration modes.
+func MoveTypesForVersionMigration(
+	configuration *config.ObjectModelConfiguration,
+) *Stage {
+	stage := NewStage(
+		MoveTypesForVersionMigrationStageID,
+		"Move types into new packages to support version migration",
+		func(ctx context.Context, state *State) (*State, error) {
+			processor := newVersionMigrationFactory(configuration, state.Definitions())
+			newDefinitions, err := processor.Process(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			return state.WithDefinitions(newDefinitions), nil
+		})
+
+	stage.RequiresPostrequisiteStages(CreateStorageTypesStageID)
+
+	return stage
+}
+
+type versionMigrationFactory struct {
+	configuration     *config.ObjectModelConfiguration
+	definitions       astmodel.TypeDefinitionSet
+	lastLegacyVersion string
+}
+
+func newVersionMigrationFactory(
+	configuration *config.ObjectModelConfiguration,
+	definitions astmodel.TypeDefinitionSet,
+) *versionMigrationFactory {
+	return &versionMigrationFactory{
+		configuration:     configuration,
+		definitions:       definitions,
+		lastLegacyVersion: "v2.16.0", //!! Change to v2.16.0 once hybrid mode is in place
+	}
+}
+
+func (p *versionMigrationFactory) Process(ctx context.Context) (astmodel.TypeDefinitionSet, error) {
+	// Find all the resources currently using legacy mode
+	// versioning that need to be moved to use new style versioning
+	toMove := p.findLegacyModeResourcesToMove()
```

Comment:
Does this mean if I set mode legacy on a resource adding in 2.17 it will use new-style versioning? Wondering if it makes sense to just block that at generation time rather than "ignore" it?

Reply:
The mode is set per `goup`, not per version. #nitpick

If a brand new group is introduced in v2.17 AND that group is configured for `legacy` mode, then Yes, that migration would happen. I've added a check to fail the generator if this happens.